plugins/janus_nosip.c

```diff
@@ -0,0 +1,2213 @@
+		JANUS_LOG(LOG_INFO, "Closed peer's audio recording %s\n", session->arc_peer->filename ? session->arc_peer->filename : "??");
+		janus_recorder_free(session->arc_peer);
+	}
+	session->arc_peer = NULL;
+	if(session->vrc) {
+		janus_recorder_close(session->vrc);
+		JANUS_LOG(LOG_INFO, "Closed user's video recording %s\n", session->vrc->filename ? session->vrc->filename : "??");
+		janus_recorder_free(session->vrc);
+	}
+	session->vrc = NULL;
+	if(session->vrc_peer) {
+		janus_recorder_close(session->vrc_peer);
+		JANUS_LOG(LOG_INFO, "Closed peer's video recording %s\n", session->vrc_peer->filename ? session->vrc_peer->filename : "??");
+		janus_recorder_free(session->vrc_peer);
+	}
+	session->vrc_peer = NULL;
+	janus_mutex_unlock(&session->rec_mutex);
+	/* FIXME Simulate a "hangup" coming from the browser */
+	janus_nosip_message *msg = g_malloc0(sizeof(janus_nosip_message));
+	msg->handle = handle;
+	msg->message = json_pack("{ss}", "request", "hangup");
+	msg->transaction = NULL;
+	msg->jsep = NULL;
+	g_async_queue_push(messages, msg);
+}
+
+/* Thread to handle incoming messages */
+static void *janus_nosip_handler(void *data) {
+	JANUS_LOG(LOG_VERB, "Joining NoSIP handler thread\n");
+	janus_nosip_message *msg = NULL;
+	int error_code = 0;
+	char error_cause[512];
+	json_t *root = NULL;
+	while(g_atomic_int_get(&initialized) && !g_atomic_int_get(&stopping)) {
+		msg = g_async_queue_pop(messages);
+		if(msg == NULL)
+			continue;
+		if(msg == &exit_message)
+			break;
+		if(msg->handle == NULL) {
+			janus_nosip_message_free(msg);
+			continue;
+		}
+		janus_nosip_session *session = NULL;
+		janus_mutex_lock(&sessions_mutex);
+		if(g_hash_table_lookup(sessions, msg->handle) != NULL ) {
+			session = (janus_nosip_session *)msg->handle->plugin_handle;
+		}
+		janus_mutex_unlock(&sessions_mutex);
+		if(!session) {
+			JANUS_LOG(LOG_ERR, "No session associated with this handle...\n");
+			janus_nosip_message_free(msg);
+			continue;
+		}
+		if(session->destroyed) {
+			janus_nosip_message_free(msg);
+			continue;
+		}
+		/* Handle request */
+		error_code = 0;
+		root = msg->message;
+		if(msg->message == NULL) {
+			JANUS_LOG(LOG_ERR, "No message??\n");
+			error_code = JANUS_NOSIP_ERROR_NO_MESSAGE;
+			g_snprintf(error_cause, 512, "%s", "No message??");
+			goto error;
+		}
+		if(!json_is_object(root)) {
+			JANUS_LOG(LOG_ERR, "JSON error: not an object\n");
+			error_code = JANUS_NOSIP_ERROR_INVALID_JSON;
+			g_snprintf(error_cause, 512, "JSON error: not an object");
+			goto error;
+		}
+		JANUS_VALIDATE_JSON_OBJECT(root, request_parameters,
+			error_code, error_cause, TRUE,
+			JANUS_NOSIP_ERROR_MISSING_ELEMENT, JANUS_NOSIP_ERROR_INVALID_ELEMENT);
+		if(error_code != 0)
+			goto error;
+		json_t *request = json_object_get(root, "request");
+		const char *request_text = json_string_value(request);
+		json_t *result = NULL, *localjsep = NULL;
+
+		if(!strcasecmp(request_text, "generate") || !strcasecmp(request_text, "process")) {
```

Comment:
consider splinting this into separate "handle" functions - handle_generate, handle_recording, etc.
maybe even go deeper and split those as well - handle_recording_start, handle_recording_stop, handle_generate_answer, etc.